src/client/meshgen/collector.cpp

```diff
@@ -0,0 +1,119 @@
+#include "collector.h"
+#include "log.h"
+#include "mesh.h"
+
+void MeshCollector::append(const TileSpec &tile, const video::S3DVertex *vertices,
+		u32 numVertices, const u16 *indices, u32 numIndices)
+{
+	for (int layernum = 0; layernum < MAX_TILE_LAYERS; layernum++) {
+		const TileLayer *layer = &tile.layers[layernum];
+		if (layer->texture_id == 0)
+			continue;
+		append(*layer, vertices, numVertices, indices, numIndices, layernum,
+				tile.world_aligned);
+	}
+}
+
+void MeshCollector::append(const TileLayer &layer, const video::S3DVertex *vertices,
+		u32 numVertices, const u16 *indices, u32 numIndices, u8 layernum,
+		bool use_scale)
+{
+	if (numIndices > 65535) {
+		dstream << "FIXME: MeshCollector::append() called with numIndices="
+			<< numIndices << " (limit 65535)" << std::endl;
+		return;
+	}
+	std::vector<PreMeshBuffer> *buffers = &prebuffers[layernum];
+
+	PreMeshBuffer *p = NULL;
+	for (PreMeshBuffer &pp : *buffers) {
+		if (pp.layer == layer && pp.indices.size() + numIndices <= 65535) {
+			p = &pp;
+			break;
+		}
+	}
+
+	if (p == NULL) {
+		PreMeshBuffer pp;
+		pp.layer = layer;
+		buffers->push_back(pp);
+		p = &(*buffers)[buffers->size() - 1];
+	}
+
+	f32 scale = 1.0;
+	if (use_scale)
+		scale = 1.0 / layer.scale;
+
+	u32 vertex_count = p->vertices.size();
+	for (u32 i = 0; i < numVertices; i++) {
+		video::S3DVertex vert(vertices[i].Pos, vertices[i].Normal,
+				vertices[i].Color, scale * vertices[i].TCoords);
+		p->vertices.push_back(vert);
+	}
+
+	for (u32 i = 0; i < numIndices; i++) {
+		u32 j = indices[i] + vertex_count;
+		p->indices.push_back(j);
+	}
+}
+
+void MeshCollector::append(const TileSpec &tile, const video::S3DVertex *vertices,
+		u32 numVertices, const u16 *indices, u32 numIndices, v3f pos,
+		video::SColor c, u8 light_source)
+{
+	for (int layernum = 0; layernum < MAX_TILE_LAYERS; layernum++) {
+		const TileLayer *layer = &tile.layers[layernum];
+		if (layer->texture_id == 0)
+			continue;
+		append(*layer, vertices, numVertices, indices, numIndices, pos, c,
+				light_source, layernum, tile.world_aligned);
+	}
+}
+
+void MeshCollector::append(const TileLayer &layer, const video::S3DVertex *vertices,
+		u32 numVertices, const u16 *indices, u32 numIndices, v3f pos,
+		video::SColor c, u8 light_source, u8 layernum, bool use_scale)
+{
+	if (numIndices > 65535) {
```

Comment:
use UINT16_MAX please for those values

Reply:
`U16_MAX`, also these lines are only moved, not modified in its contents.

Reply:
> these lines are only moved, not modified in its contents.

...and the check is actually wrong. Vertex amount is limited by the implementation, not index amount, as vertex index is `u16`. And as source indices are `u16` as well... there can’t be more than 65536 vertices anyway; so replacing with correct check.